lightningd/plugin_hook.h

```diff
@@ -0,0 +1,118 @@
+ * plugins that have registered a hook for a given event how it'd like
+ * to proceed. This allows plugins to deviate from the default
+ * behavior that `lightningd` otherwise implements.
+ *
+ * Examples include storing an additional backup of important
+ * information belonging to the wallet before committing to it, or
+ * holding an incoming payment that is guaranteed to succeed for some
+ * time in order to check that the delivery of goods works correctly,
+ * giving the option of instantly refunding should something go wrong.
+ *
+ * Hooks are commonly structured into a number of converter functions
+ * and a callback. The converter functions convert from an internal
+ * struct representation of the method arguments to a JSON-object for
+ * delivery to the plugin, and from a JSON-object to the internal
+ * representation:
+ *
+ * - `serialize_payload` which takes a payload of type `payload_type`
+ *   and serializes it into the given `json_stream`. `
+ *
+ * - `deserialize_response` takes a `json_stream` and parses it into a
+ *   new struct of type `response_type`,
+ *
+ * - `response_cb` is called once the plugin has responded and the
+ *   response has been parsed by `deserialize_response`. In addition
+ *   an arrbitrary additional argument of type `cb_arg_type` can be
+ *   passed along that may contain any additional context necessary.
+ *
+ *
+ * To make hook invocations easier, each hook registered with
+ * `REGISTER_PLUGIN_HOOK` provides a `plugin_hook_call_hookname`
+ * function that performs typechecking at compile time, and makes sure
+ * that all the provided functions for serialization, deserialization
+ * and callback have the correct type.
+ */
+
+struct plugin_hook {
+	const char *name;
+	void (*response_cb)(void *arg, void *response);
+	void (*serialize_payload)(void *src, struct json_stream *dest);
+	void *(*deserialize_response)(const tal_t *, const char *buffer,
+				      const jsmntok_t *toks);
+
+	/* Which plugin has registered this hook? */
+	struct plugin *plugin;
+};
+AUTODATA_TYPE(hooks, struct plugin_hook);
+
+/* Do not call this directly, rather use the `plugin_hook_call_name`
+ * wrappers generated by the `PLUGIN_HOOK_REGISTER` macro.
+ */
+void plugin_hook_call_(struct lightningd *ld, const struct plugin_hook *hook,
+		       void *payload, void *cb_arg);
+
+
+/* Create a small facade in from of `plugin_hook_call_` to make sure
+ * arguments are of the correct type before downcasting them to `void
+ * *`. Not really necessary, but nice since it also makes sure that
+ * the method-name is correct for the call.
+ */
+/* FIXME: Find a way to avoid back-to-back declaration and definition */
+#define PLUGIN_HOOK_CALL_DEF(name, payload_type, response_cb_arg_type)         \
+	void plugin_hook_call_##name(struct lightningd *ld,                    \
+				     payload_type payload,                     \
+				     response_cb_arg_type cb_arg);             \
```

Comment:
Yes, like mentioned above this one is the declaration of the function (no function body) and the lower one is the function definition (which contains the body). It's a bit of an antipattern since the definition should be in the matching `.c` file, but since we'd like to define this in a macro it has to be in the same place, and since we may want to call this from somewhere other than the module defining it (hooks should be defined together with the data they act on, and called from the entrypoints of the hooks), this is the best I could come up with.

Maybe @rustyrussell has a better option? An alternative that I haven't tried yet is to use a `#define` instead of this tiny function, avoiding all the linker issues altogether.

Reply:
This Just Worked for me?

```diff
diff --git a/lightningd/plugin_hook.h b/lightningd/plugin_hook.h
index 48dffe23..74be1e74 100644
--- a/lightningd/plugin_hook.h
+++ b/lightningd/plugin_hook.h
@@ -71,12 +71,8 @@ void plugin_hook_call_(struct lightningd *ld, const struct plugin_hook *hook,
  * *`. Not really necessary, but nice since it also makes sure that
  * the method-name is correct for the call.
  */
-/* FIXME: Find a way to avoid back-to-back declaration and definition */
 #define PLUGIN_HOOK_CALL_DEF(name, payload_type, response_cb_arg_type)         \
-       void plugin_hook_call_##name(struct lightningd *ld,                    \
-                                    payload_type payload,                     \
-                                    response_cb_arg_type cb_arg);             \
-       void plugin_hook_call_##name(struct lightningd *ld,                    \
+       static inline void plugin_hook_call_##name(struct lightningd *ld,      \
                                     payload_type payload,                     \
                                     response_cb_arg_type cb_arg)              \
        {                                                                      \
```

Reply:
Interesting, I thought `static inline` would only be callable in the same compilation unit, but you're right, since we are `#include`ing the file, this probably means the client header as well. Will fixup :+1: